homeassistant/components/rainmachine/__init__.py

```diff
@@ -0,0 +1,240 @@
+    TYPE_HOT_DAYS: ('Extra Water on Hot Days', 'mdi:thermometer-lines'),
+    TYPE_HOURLY: ('Hourly Restrictions', 'mdi:cancel'),
+    TYPE_MONTH: ('Month Restrictions', 'mdi:cancel'),
+    TYPE_RAINDELAY: ('Rain Delay Restrictions', 'mdi:cancel'),
+    TYPE_RAINSENSOR: ('Rain Sensor Restrictions', 'mdi:cancel'),
+    TYPE_WEEKDAY: ('Weekday Restrictions', 'mdi:cancel'),
+}
+
+SENSORS = {
+    TYPE_FREEZE_TEMP: ('Freeze Protect Temperature', 'mdi:thermometer', '°C'),
+}
+
+BINARY_SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_MONITORED_CONDITIONS):
+        vol.All(cv.ensure_list, [vol.In(BINARY_SENSORS)])
+})
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_MONITORED_CONDITIONS):
+        vol.All(cv.ensure_list, [vol.In(SENSORS)])
+})
+
+SERVICE_START_PROGRAM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PROGRAM_ID): cv.positive_int,
+})
+
+SERVICE_START_ZONE_SCHEMA = vol.Schema({
+    vol.Required(CONF_ZONE_ID): cv.positive_int,
+    vol.Optional(CONF_ZONE_RUN_TIME, default=DEFAULT_ZONE_RUN):
+        cv.positive_int,
+})
+
+SERVICE_STOP_PROGRAM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PROGRAM_ID): cv.positive_int,
+})
+
+SERVICE_STOP_ZONE_SCHEMA = vol.Schema({
+    vol.Required(CONF_ZONE_ID): cv.positive_int,
+})
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_ZONE_RUN_TIME): cv.positive_int
+})
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema({
+            vol.Required(CONF_IP_ADDRESS): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+            vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+            vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+            vol.Optional(CONF_BINARY_SENSORS): BINARY_SENSOR_SCHEMA,
+            vol.Optional(CONF_SENSORS): SENSOR_SCHEMA,
+            vol.Optional(CONF_SWITCHES): SWITCH_SCHEMA,
+        })
+    },
+    extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the RainMachine component."""
+    from regenmaschine import Authenticator, Client
+    from regenmaschine.exceptions import RainMachineError
+
+    conf = config[DOMAIN]
+    ip_address = conf[CONF_IP_ADDRESS]
+    password = conf[CONF_PASSWORD]
+    port = conf[CONF_PORT]
+    ssl = conf[CONF_SSL]
+
+    try:
+        auth = Authenticator.create_local(
+            ip_address, password, port=port, https=ssl)
+        rainmachine = RainMachine(hass, Client(auth))
+        rainmachine.update()
+        hass.data[DATA_RAINMACHINE] = rainmachine
+    except RainMachineError as exc:
+        _LOGGER.error('An error occurred: %s', str(exc))
+        hass.components.persistent_notification.create(
+            'Error: {0}<br />'
+            'You will need to restart hass after fixing.'
+            ''.format(exc),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    binary_sensor_config = conf.get(CONF_BINARY_SENSORS, {})
+    discovery.load_platform(
```

Comment:
You can do a for loop that gets the component config and loads the three platforms. Iterate over config schema type and component.